vpr/src/route/rr_graph_fpga_interchange.cpp

```diff
@@ -0,0 +1,1872 @@
+
+            device_ctx_.rr_graph_builder.emplace_back_edge(RRNodeId(src), RRNodeId(sink), 0);
+            device_ctx_.rr_graph_builder.emplace_back_edge(RRNodeId(sink), RRNodeId(src), 0);
+        }
+    }
+
+    void pack_pips() {
+        for (auto& i : pips_) {
+            int node1, node2;
+            std::tie(node1, node2) = i.first;
+            int sw_id, tile_id;
+            std::tuple<int, int, int, bool> metadata;
+            std::tie(sw_id, tile_id, metadata) = i.second;
+            VTR_ASSERT(sw_id > 1);
+            int name, wire0, wire1;
+            bool forward;
+            std::tie(name, wire0, wire1, forward) = metadata;
+            t_location loc = tile_loc_bimap_[tile_id];
+            auto virtual_key1 = virtual_redirect_[std::make_tuple(node1, loc)];
+            auto virtual_key2 = virtual_redirect_[std::make_tuple(node2, loc)];
+
+            auto key1 = std::make_tuple(std::get<0>(virtual_key1),
+                                        std::get<1>(virtual_key1),
+                                        virtual_beg_to_real_[virtual_key1]);
+            auto key2 = std::make_tuple(std::get<0>(virtual_key2),
+                                        std::get<1>(virtual_key2),
+                                        virtual_beg_to_real_[virtual_key2]);
+
+            int src, sink;
+            src = loc_type_idx_to_rr_idx_[key1];
+            sink = loc_type_idx_to_rr_idx_[key2];
+            device_ctx_.rr_graph_builder.emplace_back_edge(RRNodeId(src), RRNodeId(sink), sw_id);
+
+            // TODO: find a more efficient way to pack PIP information to build the physical netlist out
+            //       of a routed design
+            char metadata_[100];
+            char* temp = int_to_string(metadata_, name);
+            *temp++ = '.';
+            temp = int_to_string(temp, wire0);
+            *temp++ = ',';
+            temp = int_to_string(temp, wire1);
+            *temp++ = '.';
+            temp = int_to_string(temp, forward ? 1 : 0);
+            *temp++ = 0;
+
+            vpr::add_rr_edge_metadata(src, sink, sw_id, vtr::string_view("FPGAInterchange"), vtr::string_view(metadata_));
+        }
+    }
+
+    void pack_rr_edges() {
+        pack_tiles_edges();
+        pack_chans_edges();
+        pack_pips();
+        device_ctx_.rr_graph_builder.mark_edges_as_rr_switch_ids();
+        device_ctx_.rr_graph_builder.partition_edges();
+    }
+
+    /** @brief Fills the device RR graph.
+     */
+    void pack_to_rr_graph() {
+        pack_rr_nodes();
+        pack_rr_edges();
+    }
+
+    /** @brief Finilizes the RR graph generation.
+     *
+     *      - Builds the node lookup
+     *      - Inits fan_ins
+     *      - Loads indexed data
+     */
+    void finish_rr_generation() {
+        for (t_rr_type rr_type : RR_TYPES)
+            if (rr_type == CHANX)
+                device_ctx_.rr_graph_builder.node_lookup().resize_nodes(grid_.height(), grid_.width(), rr_type, NUM_SIDES);
```

Comment:
I cannot find any node registered to the ``node_lookup()``. 
This may cause router failed due to the needs on ``node_lookup().find_node()``.
Can you double check?

Some example:

https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/b7a94b90a8f6692da3f1b784e61fab8cc82187d3/vpr/src/route/rr_graph2.cpp#L1076

Reply:
As far as I know function `finish_rr_generation()` is responsible for building node_lookup when running loop in lines 1771 to 1774. Calling  `add_mode_to_all_locs(node)` should be enough to register `node` in lookup.

Reply:
Got it. Sorry for missing the details.